### ISSUE TITLE
PIP error messages when installing packages

### ISSUE THREAD

**Issue by JRSlotman** [dhopp1/nowcastLSTM#1]
Having never used Python before, I struggled a little to get the packages installed. First, PIP produced errors relating to the path length. I had to follow the steps in https://www.howtogeek.com/266621/how-to-make-windows-10-accept-file-paths-over-260-characters/ to proceed. Then, PIP produced warnings. This article (https://stackoverflow.com/questions/49966547/pip-10-0-1-warning-consider-adding-this-directory-to-path-or) helped me overcome that issue. 

In RStudio, the initialize_session() command initially produced errors. Turns out I had to update the Rcpp package (https://stackoverflow.com/questions/68416435/rcpp-package-doesnt-include-rcpp-precious-remove). Now I hope it finally works!

**Comment by dhopp1** [comment]
Thanks for sharing the hurdles you came across in your install process! I will collect people's experiences like this and add it to an eventual troubleshooting section in the README. Coming from an R background the Python install requirements can be onerous I know. Let me know if you have any other issues getting it installed, happy to help you out. If you do finally get it working on your system, go ahead and close the issue.

**Comment by JRSlotman** [comment]
I seem to have messed things up with the Python installation. I tried uninstalling and reinstalling Python, all the packages and the R package but I still get the same error: "Error in py_run_string_impl(code, local, convert) : 
  ModuleNotFoundError: No module named 'nowcast_lstm'" and the same if I try to run the other py_run_string() commands, except somehow for numpy.

The Python packages are installed in:
c:\users\jrslo\appdata\local\programs\python\python39\lib\site-packages
and Python is installed in
C:\Users\jrslo\AppData\Local\Programs\Python\Python39

The Python path and the Scripts path have also been added to the Environment Variables.

Any idea what I can do?

**Comment by dhopp1** [comment]
We can break it down into 2 steps:

1) getting the Python installation working correctly, including installation of the nowcast_lstm library
2) getting the R wrapper installed and working correctly, interfacing correctly with Python

I'm pretty sure your problem is stemming from 1), so let's get that working first before we worry about 2). 

If you followed the steps to install Python and troubleshot on your machine following the instructions of the README and that didn't end up working, I'd recommend trying to install Python via [Miniconda ](https://docs.conda.io/en/latest/miniconda.html) instead of separately. This also gives you the handy [Anaconda Prompt](https://docs.anaconda.com/anaconda/user-guide/getting-started/) which is more similar to Unix command lines. Then you can `pip install ...` everything you need from there. 

The final goal you're after to make sure 1) is working correctly is to be able to run `>python` from your command line/Anaconda prompt, then from within Python be able to run `>>>from nowcast_lstm import LSTM`. If that then works, then you know 1) is all good, and it should be pretty easy to get 2) from there. 

So give it a shot with miniconda, if that still doesn't work, shoot me an email at daniel.hopp@un.org and I'd be happy to schedule a quick call with screensharing to get it sorted.

**Comment by JRSlotman** [comment]
Alright, I got it to work! 

I followed your instructions to install Python through Miniconda and then went back to RStudio. At first, I still got the same error. I tried various locations for the Python path following the instructions here: https://stackoverflow.com/questions/52462780/how-to-properly-set-the-python-path-for-the-reticulate-package-in-osx-high-sierr. Ultimately, the path that worked was the r-miniconda installation folder, with the following line of code:
reticulate::use_python(python = 'C:/Users/jrslo/AppData/Local/r-miniconda', required = T)

Retracing my steps, I think it would've been better to avoid the Microsoft Store Python installation and go straight to Conda. Finding the right Python path is probably the most tedious aspect in the installation but I guess it's a good learning experience. I've never used Python or the reticulate package so I had to read up a little on both to understand the synergy and identify where things failed. Despite the struggle, I think the package is wonderful and I congratulate you for taking the effort to provide this R wrapper. 

Thank you so much for your help! I hope others who try to use your package face less issues or can use the discussion here to solve them.

**Comment by dhopp1** [comment]
That's great, as I said I'll collect these and others' experiences and incorporate them into a troubleshooting section of the README. If there are any questions on the actual modelling or data work, also feel free to reach out directly or here.

**Comment by JRSlotman** [comment]
Just sharing another update on the installation, after successfully having set up the package on another machine in considerably less time. I found that installing Python/Miniconda and all the packages through the reticulate functions avoids all system environment issues. Specifically, I used the following code:
`install_miniconda(path = miniconda_path(), update = TRUE, force = FALSE)`
`conda_install(miniconda_path(), "dill numpy pandas pmdarima torch nowcast-lstm", pip = TRUE)`
You can then also substitute miniconda_path() for Python path in the use_python() and initialize_session() functions. Hope this helps any future users!

**Comment by dhopp1** [comment]
That's awesome Julian, much more elegant than the current instructions, keeping everything within R. Thanks so much for sharing. I will recreate this on another machine and then update the README instructions accordingly.

**Comment by dhopp1** [comment]
@JRSlotman Did you come up against the error of `No module named pip` when trying to run `conda_install(miniconda_path(), "dill numpy pandas pmdarima torch nowcast-lstm", pip = TRUE)`? I tried to get around it by running `py_install()` instead, but it kills the torch installation for some reason.

**Comment by JRSlotman** [comment]
Admittedly, I summarized things a little in the code I put on Github, (possibly incorrectly) assuming that it would work the same as what I actually did. When I managed to get it set up all through reticulate functions, I actually first ran three separate install lines similar to your own instructions: First, on "dill numpy pandas pmdarim", then on "torch" and then on "nowcast-lstm". The options were all the same, i.e.
`conda_install(miniconda_path(), "packages", pip = TRUE)`. 

I'd like to try again but I've run out of machines and I don't want to uninstall everything now.

**Comment by dhopp1** [comment]
Thanks I actually figured it out, `py_install(conda=miniconda_path(), "dill numpy pandas pmdarima torch nowcast-lstm", pip=TRUE)` works perfectly fine and doesn't have the pip issue. The issue was I was running everything on a Docker container with only 2gb of RAM, and torch specifically is around 800mb-1gb, so when installing with pip it sits in RAM before installing and maxed out the container's memory (hence the "killed" error). I upped the RAM on the container and the problem went away and everything works as expected with your approach. I'm updating the README now.
